### ISSUE TITLE
给各位探探坑，讲下自己遇到并解决的问题

### ISSUE THREAD

**Issue by bigKoki** [DetectionTeamUCAS/RetinaNet_Tensorflow_Rotation#8]
①tfrecord读取问题：
按照作者的方法生成tfrecord文件，首先要裁剪图片，生成xml文件，但是由于有的txt文件在坐标数据上面多了两行不需要的文字，这会导致生成xml文件出错，可以在代码中略过这两行。
②训练的时候提示shufflebatch=0：
网上百度的结果都是说什么未初始化，我从next_batch代码中分析，应该是未正确从tfrecord文件中读取图片和标签，我调试了一下，发现是相对路径无法找到，于是我改成了绝对路径就可以了。
③预训练模型问题：
要将完整的res101.ckpt，resnet101_v1d.ckpt.index.......等等这些文件全放入预训练文件夹
④找不到bbox啥函数或者文件：
要根据作者的方法编译setup文件，在win10是无法编译的。

**Comment by yangxue0827** [comment]
请问你训练出效果了吗？ @864546664

**Comment by bigKoki** [comment]
> 请问你训练出效果了吗？ @864546664

正在训练呢，今天中午才弄好，估计要下午才能看结果

**Comment by yangxue0827** [comment]
嗯，希望到时候反馈一下，因为我看到我这人说nan @864546664

**Comment by bigKoki** [comment]
> 嗯，希望到时候反馈一下，因为我看到我这人说nan @864546664

我的loss值比较正常，不过训练1000轮了，loss还在1.7左右徘徊，刚开始也是1.7

**Comment by yangxue0827** [comment]
能贴一下你的cfgs.py吗 @864546664

**Comment by bigKoki** [comment]
> 能贴一下你的cfgs.py吗 @864546664


麻烦作者大大了
# ------------------------------------------------
VERSION = 'RetinaNet_DOTA_1x_20190606'
NET_NAME = 'resnet101_v1d'  # 'MobilenetV2'
ADD_BOX_IN_TENSORBOARD = True

# ---------------------------------------- System_config
ROOT_PATH = os.path.abspath('../')
print(20*"++--")
print(ROOT_PATH)
GPU_GROUP = "0"
NUM_GPU = len(GPU_GROUP.strip().split(','))
SHOW_TRAIN_INFO_INTE = 20
SMRY_ITER = 200
SAVE_WEIGHTS_INTE = 27000

SUMMARY_PATH = ROOT_PATH + '/output/summary'
TEST_SAVE_PATH = ROOT_PATH + '/tools/test_result'

if NET_NAME.startswith("resnet"):
    weights_name = NET_NAME
elif NET_NAME.startswith("MobilenetV2"):
    weights_name = "mobilenet/mobilenet_v2_1.0_224"
else:
    raise Exception('net name must in [resnet_v1_101, resnet_v1_50, MobilenetV2]')

PRETRAINED_CKPT = ROOT_PATH + '/data/pretrained_weights/' + weights_name + '.ckpt'
TRAINED_CKPT = os.path.join(ROOT_PATH, 'output/trained_weights')
EVALUATE_DIR = ROOT_PATH + '/output/evaluate_result_pickle/'

# ------------------------------------------ Train config
RESTORE_FROM_RPN = False
FIXED_BLOCKS = 1  # allow 0~3
FREEZE_BLOCKS = [True, False, False, False, False]  # for gluoncv backbone
USE_07_METRIC = True

MUTILPY_BIAS_GRADIENT = 2.0  # if None, will not multipy
GRADIENT_CLIPPING_BY_NORM = 10.0  # if None, will not clip

CLS_WEIGHT = 1.0
REG_WEIGHT = 1.0
USE_IOU_FACTOR = True

BATCH_SIZE = 1
EPSILON = 1e-5
MOMENTUM = 0.9
LR = 5e-4  # * NUM_GPU * BATCH_SIZE
DECAY_STEP = [SAVE_WEIGHTS_INTE*12, SAVE_WEIGHTS_INTE*16, SAVE_WEIGHTS_INTE*20]
MAX_ITERATION = SAVE_WEIGHTS_INTE*20
WARM_SETP = int(1.0 / 8.0 * SAVE_WEIGHTS_INTE)

# -------------------------------------------- Data_preprocess_config
DATASET_NAME = 'DOTA'  # 'pascal', 'coco'
PIXEL_MEAN = [123.68, 116.779, 103.939]  # R, G, B. In tf, channel is RGB. In openCV, channel is BGR
PIXEL_MEAN_ = [0.485, 0.456, 0.406]
PIXEL_STD = [0.229, 0.224, 0.225]  # R, G, B. In tf, channel is RGB. In openCV, channel is BGR
IMG_SHORT_SIDE_LEN = 800
IMG_MAX_LENGTH = 800
CLASS_NUM = 15

# --------------------------------------------- Network_config
SUBNETS_WEIGHTS_INITIALIZER = tf.random_normal_initializer(mean=0.0, stddev=0.01, seed=None)
SUBNETS_BIAS_INITIALIZER = tf.constant_initializer(value=0.0)
PROBABILITY = 0.01
FINAL_CONV_BIAS_INITIALIZER = tf.constant_initializer(value=-math.log((1.0 - PROBABILITY) / PROBABILITY))
WEIGHT_DECAY = 1e-4

# ---------------------------------------------Anchor config
LEVEL = ['P3', 'P4', 'P5', 'P6', 'P7']
BASE_ANCHOR_SIZE_LIST = [32, 64, 128, 256, 512]
ANCHOR_STRIDE = [8, 16, 32, 64, 128]
ANCHOR_SCALES = [2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)]
ANCHOR_RATIOS = [1, 1 / 3., 3., 5., 1 / 5.]
ANCHOR_ANGLES = [-90, -75, -60, -45, -30, -15]
ANCHOR_SCALE_FACTORS = None
USE_CENTER_OFFSET = True
METHOD = 'R'
USE_ANGLE_COND = False

# --------------------------------------------RPN config
SHARE_NET = True
USE_P5 = True
IOU_POSITIVE_THRESHOLD = 0.5
IOU_NEGATIVE_THRESHOLD = 0.4

NMS = True
NMS_IOU_THRESHOLD = 0.5
MAXIMUM_DETECTIONS = 100
FILTERED_SCORE = 0.05
VIS_SCORE = 0.4

# --------------------------------------------NAS FPN config
NUM_FPN = 0
NUM_NAS_FPN = 0
USE_RELU = True
FPN_CHANNEL = 256

**Comment by yangxue0827** [comment]
这个训练1000轮还太少，需要先训练几个epoch才会有明显的变化，tensorboard里才会陆续出现检测框。 @864546664

**Comment by bigKoki** [comment]
> 这个训练1000轮还太少，需要先训练几个epoch才会有明显的变化，tensorboard里才会陆续出现检测框。 @864546664

好的，我下午再看看结果

**Comment by yangxue0827** [comment]
我更新了代码，如果你跑着NAN了，可以更新一下。建议先跑通一个默认的配置文件。 @864546664

**Comment by bigKoki** [comment]
> 我更新了代码，如果你跑着NAN了，可以更新一下。建议先跑通一个默认的配置文件。 @864546664
作者您好，我现在正在训练，大概训练了1400step，我在tensorboaed里看训练过程，gtboxes_h_gpu_0图像和gtboxes_r_gpu_0图像都是有框的，而/final_detection_gpu_0图像没有框，请问这正常吗？在问一下，大概要多少step才能达到一个好的效果呢？

**Comment by yangxue0827** [comment]
两三个epoch，你训练速度怎么这么慢，没有使用GPU吗? @864546664

**Comment by bigKoki** [comment]
> 两三个epoch，你训练速度怎么这么慢，没有使用GPU吗? @864546664

貌似是的，没有使用gpu,我现在loss也变成nan了,待会去试试你的新代码

**Comment by bigKoki** [comment]
> 两三个epoch，你训练速度怎么这么慢，没有使用GPU吗? @864546664
使用了新代码，之前的一些问题也没有了，现在训练了240k了，效果有明显的改善

**Comment by yangxue0827** [comment]
感谢反馈，还有什么问题可以随时留言 @864546664

**Comment by yangxue0827** [comment]
请问你进行提交测试了吗，效果如何 @864546664

**Comment by bigKoki** [comment]
> 请问你进行提交测试了吗，效果如何 @864546664

我还没有训练好，现在的total loss仍在0.7左右，但是我去测试了一下，map在0.19左右，因为我测试的数据集有18个类，所以会有影响，我等明天训练完了，再去测试一次

**Comment by yangxue0827** [comment]
你不是训练DOTA吗？ @864546664

**Comment by bigKoki** [comment]
> 你不是训练DOTA吗？ @864546664

差不多，但是有18个类

**Comment by yangxue0827** [comment]
是最近那个比赛是吗？ @864546664
 那0.19是正常， 什么东西都不加的话

**Comment by bigKoki** [comment]
> 是最近那个比赛是吗？ @864546664
> 那0.19是正常， 什么东西都不加的话

是啊，我只是帮别人跑跑代码，测试效果

**Comment by yangxue0827** [comment]
那你们有跑过我的R2CNN吗，在那个数据集的效果大概多少？ @864546664

**Comment by bigKoki** [comment]
> 那你们有跑过我的R2CNN吗，在那个数据集的效果大概多少？ @864546664

还没有呢，队长准备借鉴你的方法做。请问你的r2cnn和这个的技术，哪个更好实现呢，如果准确率差不多的话，我们想研究实现更快的

**Comment by yangxue0827** [comment]
dota的话差不多，速度这个更快，当然你要打比赛的话肯定推荐双阶段方法，能改进的地方多。 @864546664

**Comment by bigKoki** [comment]
> dota的话差不多，速度这个更快，当然你要打比赛的话肯定推荐双阶段方法，能改进的地方多。 @864546664

请问你的这个方法有论文吗，能否观摩一下

**Comment by qzq-hit** [comment]
> > 是最近那个比赛是吗？ @864546664
> > 那0.19是正常， 什么东西都不加的话
> 
> 是啊，我只是帮别人跑跑代码，测试效果

你好 我也是用这个方法测试的这个比赛的数据集，请问你在测试的时候有没有出现没有测试结果的情况呢 可以给些改正建议吗？不胜感激

**Comment by bigKoki** [comment]
> > > 是最近那个比赛是吗？ @864546664
> > > 那0.19是正常， 什么东西都不加的话
> > 
> > 
> > 是啊，我只是帮别人跑跑代码，测试效果
> 
> 你好 我也是用这个方法测试的这个比赛的数据集，请问你在测试的时候有没有出现没有测试结果的情况呢 可以给些改正建议吗？不胜感激

不好意思哦，我也不是很懂，正在学习中。你跑了多少轮？我是跑了大概100k以上后测试才有效果的，前面的loss还没真正降下去，可以在tensorboard看训练的过程

**Comment by yangxue0827** [comment]
参考：
[R3Det](https://arxiv.org/abs/1908.05612)
[SCRDet](https://arxiv.org/abs/1811.07126)
@864546664

**Comment by hehu56** [comment]
@864546664 你好请问能贴一下win10下的setup文件吗
